### PR TITLE
Allow discovering test methods from base class in different assembly

### PR DESCRIPTION
**Pull request by ajryan**
Fixes issue #23

### CODE REVIEW

File: test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs [hunk 1]
```diff
@@ -101,13 +103,12 @@ public void GetTestsShouldReturnBaseTestMethodsInSameAssembly()
 
             var tests = typeEnumerator.Enumerate(out this.warnings);
 
-            var methodCount = typeof(DummyDerivedTestClass).GetMethods(BindingFlags.Instance | BindingFlags.Public).Count(m => m.DeclaringType.Assembly == typeof(DummyDerivedTestClass).Assembly);
             Assert.IsNotNull(tests);
-            Assert.AreEqual(methodCount, tests.Count);
+            Assert.IsTrue(tests.Any(t => t.TestMethod.Name == "BaseTestMethod"));
```
Comment (AbhitejJohn) [inline test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs]: >Assert.IsTrue(tests.Any(t => t.TestMethod.Name == "BaseTestMethod")); [](start = 12, length = 69)

Do add a message on assertion failure so it is easy to identify why this test failed.

File: test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs [hunk 2]
```diff
@@ -121,6 +122,58 @@ public void GetTestsShouldNotReturnBaseTestMethodsFromAnotherAssembly()
...
+            mockRunContext.Setup(dc => dc.RunSettings).Returns(mockRunSettings.Object);
+            mockRunSettings.Setup(rs => rs.SettingsXml).Returns(runSettingxml);
+
+            MSTestSettings.PopulateSettings(mockRunContext.Object);
+            this.SetupTestClassAndTestMethods(isValidTestClass: true, isValidTestMethod: true);
+            TypeEnumerator typeEnumerator = this.GetTypeEnumeratorInstance(typeof(DummyDerivedFromRemoteTestClass), Assembly.GetExecutingAssembly().FullName);
+
+            var tests = typeEnumerator.Enumerate(out this.warnings);
+
+            Assert.IsTrue(tests.All(t => t.TestMethod.Name != "BaseTestMethod"));
```
Comment (AbhitejJohn) [inline test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs]: >t.TestMethod.Name != "BaseTestMethod" [](start = 41, length = 37)

add a contains check instead.

File: src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs [hunk 3]
```diff
@@ -97,8 +97,10 @@ internal Collection<UnitTestElement> GetTests(ICollection<string> warnings)
             // Test class is already valid. Verify methods.
             foreach (var method in this.type.GetRuntimeMethods())
             {
-                // Todo: Provide settings to allow users to pick up tests from other assemblies as well.
-                if (!method.DeclaringType.GetTypeInfo().Assembly.Equals(this.type.GetTypeInfo().Assembly))
+                var isMethodDeclaredInTestTypeAssembly = method.DeclaringType.GetTypeInfo().Assembly.Equals(this.type.GetTypeInfo().Assembly);
```
Comment (AbhitejJohn) [inline src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs]: >method.DeclaringType.GetTypeInfo().Assembly.Equals(this.type.GetTypeInfo().Assembly) [](start = 57, length = 84)

This seems to be causing issues in testability. How about putting this in a separate API in reflecthelper and mocking that out in your tests?

File: src/Adapter/MSTest.CoreAdapter/MSTestSettings.cs [hunk 4]
```diff
@@ -238,6 +245,16 @@ private static MSTestSettings ToSettings(XmlReader reader)
                                 break;
                             }
 
+                        case "ENABLEBASECLASSTESTMETHODSFROMOTHERASSEMBLIES":
```
Comment (AbhitejJohn) [inline src/Adapter/MSTest.CoreAdapter/MSTestSettings.cs:248]: >ENABLEBASECLASSTESTMETHODSFROMOTHERASSEMBLIES [](start = 30, length = 45)

can we add tests for this too please? you would find them under MSTest.CoreAdapter.Unit.Tests -> MSTestSettingsTests.cs

**Comment by AbhitejJohn** [comment]
Thanks a bunch for getting this fixed quickly @ajryan. Had a few comments. Will push this in as soon as they are resolved.

**Comment by AbhitejJohn** [comment]
@pvlakshm should this be turned on by default(searching for test methods in a base class defined in another assembly) given that this is a common scenario?

File: test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs (hunk 1 above)
Comment (ajryan) [reply test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs]: Will do for all test methods I've changed/added.

File: src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs (hunk 3 above)
Comment (ajryan) [reply src/Adapter/MSTest.CoreAdapter/Discovery/TypeEnumerator.cs]: Great idea, will do.

File: src/Adapter/MSTest.CoreAdapter/MSTestSettings.cs (hunk 4 above)
Comment (ajryan) [reply src/Adapter/MSTest.CoreAdapter/MSTestSettings.cs:248]: Will do.

File: test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs (hunk 2 above)
Comment (ajryan) [reply test/UnitTests/MSTest.CoreAdapter.Unit.Tests/Discovery/TypeEnumeratorTests.cs]: How would that work? tests.Contains would expect an instance of UnitTestElement and that doesn't implement equality members, nor does TestMethod...

**Comment by ajryan** [comment]
Thanks for the quick review! I'll have a revision soon. Please see my question above about the Contains thing too...

**Comment by ajryan** [comment]
This should be ready except for the comment about using Contains

**Comment by AbhitejJohn** [comment]
@ajryan : I meant contains for the name "BaseTestMethod". Thinking again, I think we already do this in multiple other places. This shouldn't be a problem. Merging this change.
Thanks for the contribution :)
/cc @sbaid @pvlakshm

**Comment by ajryan** [comment]
Awesome, thanks for reviewing and merging so quickly! :tada:

**Comment by AbhitejJohn** [comment]
Pleasure. Thanks for the contribution. This was a long standing issue with around 96 votes on [uservoice](https://visualstudio.uservoice.com/forums/330519-team-services/suggestions/6030736-support-test-inheritance-for-base-classes-in-differ). Great impact 🎉

**Pull request merged**